File: records/19x19x19/verify_19x19x19_rank3987_gf2.py
#!/usr/bin/env python3
"""Sparse exhaustive verifier for decimal-mask GF(2) matmul certificates."""
import argparse, re, sys
from pathlib import Path

def parse(path, dims):
    lines=Path(path).read_text().splitlines(); terms=[]; declared=None
    for line in lines:
        if dims is None:
            m=re.search(r'<(\d+)\s*,\s*(\d+)\s*,\s*(\d+)>',line)
            if m: dims=tuple(map(int,m.groups()))
        f=line.split()
        if not f or f[0].startswith('#'): continue
        if f[0]=='R': f=f[1:]
        if len(f)==1 and f[0].isdigit() and declared is None: declared=int(f[0]);continue
        if len(f)==3 and all(x.isdigit() for x in f):terms.append(tuple(map(int,f)));continue
        raise ValueError(f'bad line: {line[:100]}')
    if dims is None: raise ValueError('dimensions absent; pass --dims N M P')
    if declared is not None and declared != len(terms):raise ValueError(f'declared {declared}, parsed {len(terms)}')
    return dims,terms

def bits(x):
    while x:
        q=x&-x;yield q.bit_length()-1;x^=q

def verify(dims,terms):
    n,m,p=dims; limits=(n*m,m*p,n*p); residual=set(); toggles=0
    seen=set()
    for ti,t in enumerate(terms):
        if t in seen: raise ValueError(f'duplicate term {ti}')
        seen.add(t)
        for x,b in zip(t,limits):
            if not x or x.bit_length()>b:raise ValueError(f'term {ti}: invalid factor')
        ub=list(bits(t[0]));vb=list(bits(t[1]));wb=list(bits(t[2]))
        toggles += len(ub)*len(vb)*len(wb)
        for u in ub:
          for v in vb:
            base=(u*limits[1]+v)*limits[2]
            for w in wb:
              q=base+w
              if q in residual:residual.remove(q)
              else:residual.add(q)
    for i in range(n):
      for j in range(m):
        for k in range(p):
          q=((i*m+j)*limits[1]+(j*p+k))*limits[2]+i*p+k
          if q in residual:residual.remove(q)
          else:residual.add(q)
    return toggles,len(residual)

def main():
    ap=argparse.ArgumentParser();ap.add_argument('certificate');ap.add_argument('--dims',nargs=3,type=int)
    a=ap.parse_args()
    try:
      d,t=parse(a.certificate,tuple(a.dims) if a.dims else None);tog,res=verify(d,t)
    except (OSError,ValueError) as e: print(f'FAIL: {e}',file=sys.stderr);return 1
    print(f'PASS dims=<{d[0]},{d[1]},{d[2]}> rank={len(t)} sparse_toggles={tog} target_coefficients={d[0]*d[1]*d[2]} residual={res}')
    return 0 if res==0 else 1
if __name__=='__main__':raise SystemExit(main())

File: records/19x19x19/verify_19x19x19_rank3987_gf2.cpp
#include <fstream>
#include <cstdint>
#include <iostream>
#include <sstream>
#include <string>
#include <unordered_set>
#include <vector>
using Big=std::vector<uint64_t>;
struct Term { Big u,v,w; std::string us,vs,ws; };
static Big decimal(const std::string&s){Big a(1);for(char ch:s){if(ch<'0'||ch>'9')throw 1;unsigned __int128 carry=ch-'0';for(auto&x:a){carry+=(unsigned __int128)x*10;x=(uint64_t)carry;carry>>=64;}if(carry)a.push_back((uint64_t)carry);}while(a.size()>1&&!a.back())a.pop_back();return a;}
static std::vector<int> bits(const Big&x) { std::vector<int> r;for(size_t k=0;k<x.size();k++){uint64_t z=x[k];while(z){int b=__builtin_ctzll(z);r.push_back(int(k*64+b));z&=z-1;}}return r; }
static bool zero(const Big&x){return x.size()==1&&x[0]==0;}
static int top(const Big&x){auto b=bits(x);return b.empty()?-1:b.back();}
int main(int argc,char**argv){
  if(argc!=2){std::cerr<<"usage: verify_19x19x19_rank3987_gf2 FILE\n";return 2;}
  int n=19,m=19,p=19;
  std::ifstream in(argv[1]); if(!in){std::cerr<<"FAIL: open\n";return 1;}
  std::vector<Term> ts; std::string line; long declared=-1;
  while(std::getline(in,line)){if(line.empty()||line[0]=='#')continue;std::istringstream s(line);std::string a,b,c,d;s>>a;
    if(a=="R")s>>a>>b>>c;else if(!(s>>b)){if(declared<0)declared=std::stol(a);continue;}else s>>c;
    if(a.empty()||b.empty()||c.empty()||(s>>d)){std::cerr<<"FAIL: parse\n";return 1;}try{ts.push_back({decimal(a),decimal(b),decimal(c),a,b,c});}catch(...){std::cerr<<"FAIL: decimal\n";return 1;}
  }
  if(declared>=0&&declared!=(long)ts.size()){std::cerr<<"FAIL: rank header\n";return 1;}
  const int ub=n*m,vb=m*p,wb=n*p;std::unordered_set<uint64_t> r;uint64_t toggles=0;
  std::unordered_set<std::string> seen;
  for(auto&t:ts){if(zero(t.u)||zero(t.v)||zero(t.w)||top(t.u)>=ub||top(t.v)>=vb||top(t.w)>=wb){std::cerr<<"FAIL: factor bounds\n";return 1;}
    std::string key=t.us+" "+t.vs+" "+t.ws;if(!seen.insert(key).second){std::cerr<<"FAIL: duplicate\n";return 1;}
    auto us=bits(t.u),vs=bits(t.v),ws=bits(t.w);toggles+=uint64_t(us.size())*vs.size()*ws.size();
    for(int u:us)for(int v:vs)for(int w:ws){uint64_t q=(uint64_t(u)*vb+v)*wb+w;if(!r.erase(q))r.insert(q);}
  }
  for(int i=0;i<n;i++)for(int j=0;j<m;j++)for(int k=0;k<p;k++){uint64_t q=(uint64_t(i*m+j)*vb+j*p+k)*wb+i*p+k;if(!r.erase(q))r.insert(q);}
  std::cout<<(r.empty()?"PASS":"FAIL")<<" dims=<"<<n<<","<<m<<","<<p<<"> rank="<<ts.size()<<" sparse_toggles="<<toggles<<" target_coefficients="<<n*m*p<<" residual="<<r.size()<<"\n";return r.empty()?0:1;
}
